notes/overriding.md

```diff
@@ -0,0 +1,797 @@
+objects have access to the previous stage of extension via `super`. However, as
+in the NixOS module system, it is possible to address this issue using custom
+merge functions. This is a bit less expressive, but in a good way: it forces the
+merge strategy to be uniform along each field, while mechanisms like overlays or
+inheritance can do pretty much what they want.
+
+## Proposal overview
+
+We basically propose to adopt the same kind of merge, priority, and custom merge
+function-based scheme as the NixOS module system. One big difference with the
+NixOS module system is that it would be built-in in the language, being usable
+for any native record whatsoever, and having good interaction with the rest of
+the language features. The implementation would have first class access to
+locations, to the AST, the memory layout, the metadata, and so on.
+
+We first describe the ideas at a high-level. Then, we review the issue and
+challenges of this approach, and propose concrete solutions to overcome or
+mitigate them. Finally, technical operational semantics are laid out in a
+dedicated section.
+
+### Recursive records & merging
+
+**This section defines the semantics, rather than an actual efficient
+implementation**.
+
+As before, it is useful to see recursive records represented as functions - or
+constructors - that take a `self` parameter and return a final non recursive record,
+in the same way as the original overriding mechanism of Nixpkgs (the non-existing
+syntax `def := value` is used to insist on the fact that we are defining new objects):
+
+```
+r = {
+  a = 1;
+  b = a + 1;
+}
+// Definition of the representation of r
+repr(r) := fun self => {
+  a = 1;
+  b = self.a + 1;
+}
+```
+
+Field access amounts to compute a fix-point:
+
+```
+repr(r).foo := let fix = repr(r) fix in fix.foo
+```
+
+Merging simply merges the underlying representations:
+
+```
+repr(r1) & repr(r2) := fun self => r1 self & r2 self
+```
+
+### Priorities
+
+Merge is fundamentally commutative. The problem is, not all fields should be
+treated the same: some are default values that ought to be overridden, some are
+high-priority values that ought to override. The solution is to use a priority
+system, encoding the precedence of each value while retaining commutativity
+(only the priority annotation counts, not the order in which values are merged).
+
+Priorities have drawbacks. One is non-locality: the final result depends on
+other priorities of the field's values you are being merged with, potentially
+defined elsewhere. However, this can be mitigated by good defaults and an
+adapted set of priorities, such as a bottom `default`, a top `force`, and a
+infinite range of integers priorities in between. Doing so, it's easy to just
+use `default` or `force` to override or be overridden by anything, without
+knowing the precise integer priority. Integer priorities still gives freedom
+with an infinite supply of levels if required.
```

Comment:
This adds a problem of "what arbitrary integer priority do I assign this? The only way I can think to figure this out is look at everything else that sets that value, and figure out exactly what you want to override and be overridden by, and if people didn't leave enough space between their integers, you have to adjust their values. Maybe this is the best we can do, but it doesn't seem very ergonomic to me.

Maybe this has already been addressed in the NixOS module system? Just reading this document it's not clear to me how to decide on an integer.

Reply:
That's currently the biggest complaint I have with the priority system. One solution is to only allow very few and natural priorities (normal, default and force e.g.), but it seems this would be too limiting for Nix. Maybe @edolstra , @Infinisil or @regnat have more to say on the practicality of priorities.

Reply:
If we need only a few more, we could go with default < weak < normal < strong < force, but this seems like a fundamental problem of assigning these kinds of priorities. Unless there are very clear tiers of priority that naturally arise from the problem space, what you're really doing is putting an ordering on a bunch of things that are spread out across the code base.

Another example of this kind of thing is haskell's infix priorities. These work alright for a few different things `($)`, which is a "default" kind of thing, and PEMDAS operations, which have a clearly defined ordering which has to be drilled into your head as a child. I guess in general it works for multiplcation / addition type things where one operation distributes over another. In general though, if I have two random operations? I put in parentheses, because I don't have an intuition for what should be going on.